core/src/main/java/io/questdb/cutlass/line/tcp/DelegatingTlsChannel.java

```diff
@@ -0,0 +1,435 @@
+
+        int receive = delegate.receive(adjustedPtr, remainingLen);
+        if (receive < 0) {
+            throw new LineSenderException("connection closed");
+        }
+        unwrapInputBuffer.position(unwrapInputBuffer.position() + receive);
+    }
+
+    @Override
+    public int receive(long ptr, int len) {
+        try {
+            unwrapLoop();
+            unwrapOutputBuffer.flip();
+            int i = unwrapOutputBufferToPtr(ptr, len);
+            unwrapOutputBuffer.compact();
+            return i;
+        } catch (SSLException e) {
+            throw new LineSenderException("error while receiving data from questdb server", e);
+        }
+    }
+
+    private int unwrapOutputBufferToPtr(long dstPtr, int dstLen) {
+        int oldPosition = unwrapOutputBuffer.position();
+
+        assert Unsafe.getUnsafe().getLong(unwrapOutputBufferPtr, ADDRESS_FIELD_OFFSET) == unwrapOutputBufferPtr;
+        long srcPtr = unwrapOutputBufferPtr + oldPosition;
+        int srcLen = unwrapOutputBuffer.remaining();
+        int len = Math.min(dstLen, srcLen);
+        Vect.memcpy(dstPtr, srcPtr, len);
+        unwrapOutputBuffer.position(oldPosition + len);
+        return len;
+    }
+
+    private void growWrapOutputBuffer() {
+        wrapOutputBufferPtr = expandBuffer(wrapOutputBuffer, wrapOutputBufferPtr);
+    }
+
+    private void growUnwrapOutputBuffer() {
+        unwrapOutputBufferPtr = expandBuffer(unwrapOutputBuffer, unwrapOutputBufferPtr);
+    }
+
+    private void growUnwrapInputBuffer() {
+        unwrapInputBufferPtr = expandBuffer(unwrapInputBuffer, unwrapInputBufferPtr);
+    }
+
+    private static long expandBuffer(ByteBuffer buffer, long oldAddress) {
+        int oldCapacity = buffer.capacity();
+        int newCapacity = oldCapacity * 2;
+        long newAddress = Unsafe.realloc(oldAddress, oldCapacity, newCapacity, MemoryTag.NATIVE_DEFAULT);
+        resetBufferToPointer(buffer, newAddress, newCapacity);
+        return newAddress;
+    }
+
+    private static long allocateMemoryAndResetBuffer(ByteBuffer buffer, int capacity) {
+        long newAddress = Unsafe.malloc(capacity, MemoryTag.NATIVE_DEFAULT);
+        resetBufferToPointer(buffer, newAddress, capacity);
+        return newAddress;
+    }
+
+    private static void resetBufferToPointer(ByteBuffer buffer, long ptr, int len) {
+        assert buffer.isDirect();
+        Unsafe.getUnsafe().putLong(buffer, ADDRESS_FIELD_OFFSET, ptr);
+        Unsafe.getUnsafe().putLong(buffer, LIMIT_FIELD_OFFSET, len);
+        Unsafe.getUnsafe().putLong(buffer, CAPACITY_FIELD_OFFSET, len);
+    }
+
+    @Override
+    public int errno() {
+        return delegate.errno();
+    }
+
+    public void close0(boolean closeDelegate) throws IOException {
+        int prevState = state;
+        state = CLOSING;
+        if (prevState == AFTER_HANDSHAKE) {
+            try {
+                sslEngine.closeOutbound();
```

Comment:
this looks bizarre, we're are closing outbound before we attempt to send data to server ? I would also at least log error, rather than silence it. The latter hides problems.

Reply:
indeed, it does look bizarre. 

`closeOutbound()` instructs SSLEngine to produce _TLS close_notify_ alert which we then send over the wire to a server. I understood it's part of the TLS protocol.

Reply:
what I meant is that we should perhaps flush buffers before closing outbound ? There are two lines below this one that send to server after connection is closed

Reply:
flushing buffers before closing a Sender is a no-brainer. I feel ashamed for not catching this myself. Thank you!